### DISCUSSION TITLE
macro debugging

### DISCUSSION THREAD

**Discussion by 3gx** [hylang/hy#2369] [Q&A]
Is there a way to debug what is going on w/o putting `print` statements.

```hy
$ cat repro.hy
(eval-when-compile (defn foo-s [x] (+ x "str"))) ; << error here

(defmacro foo-m [x](foo_s x))

(setv foo-v (foo-m 42))

$ hy2py repro.hy
hy.errors.HyMacroExpansionError:
  File "repro.hy", line 5
    (setv foo-v (foo-m 42))
                ^--------^
expanding macro foo-m
  TypeError: unsupported operand type(s) for +: 'Integer' and 'str'
```

the issue comes from `foo-s` trying to add `x` to `"str"`, but in a more complex function it is hard to debug what is going on w/o using `print` statement. That makes writing staging programs somewhat impractical.

 Is there a way to get `hy` to print where error comes from, or perhaps a pointer where I can look at `hy` source to try to make it possible? Thanks

**Comment by Kodiologist** [comment]
It sounds like you're asking for the rest of the stack trace that would be produced in an equivalent situation without a macro. I don't think there's a way to get this as Hy stands, but this is one of several reasons that it's a good idea to put complex macro logic into a separate function, similar to how in this case you gave the `(+ …)` form to `foo-s` rather than `foo-m`. Then you can debug `foo-s` like an ordinary function, because that's what it is.

> That makes writing staging programs somewhat impractical.

What's a staging program?

> or perhaps a pointer where I can look at `hy` source to try to make it possible

You'll probably be interested in `hy/errors.py` and `hy/macros.py`.

**Reply by 3gx** [reply]
Thanks for pointers and quick response. I'll see if i can help here.

See Fig. 1 in this paper http://biboudis.github.io/papers/pcp-gpce18.pdf

Here is an example of staged factorial in `hy`:

```hy
$ cat fac.hy
(eval-and-compile
  (defn fac_s [x]
    (setv x1 (hy.eval x))
    (if (= x1 1)
      `1
      `(* ~(fac_s `(- ~x 1)) ~x1))))

(defmacro fac_m [x]
  (fac_s x))

(setv fac_v (fac_m 10))

$ hy2py fac.hy
import hy

def fac_s(x):
    x1 = hy.eval(x)
    return hy.models.Integer(1) if x1 == 1 else hy.models.Expression([hy.models.Symbol('*'), fac_s(hy.models.Expression([hy.models.Symbol('-'), x, hy.models.Integer(1)])), x1])
hy.macros.macro('fac_m')(lambda hyx_XampersandXcompiler, x: fac_s(x))
fac_v = 1 * 2 * 3 * 4 * 5 * 6 * 7 * 8 * 9 * 10
```

It is kind of cumbersome to test `fac_s` at runtime, e.g. I can execute `(fac_s '10)` at runtime, and deal with errors if there are any.

But it will make fluid experience if we get similar stack trace during macro expansion, since this is where these staged function will  mostly be used.

**Reply by 3gx** [reply]
@Kodiologist Actually, I just noticed, I do get diagnostics of what is going on, and some other stuff, with the following

```
$ HY_DEBUG=1 hy2py repro1.hy
 ...
  File "repro1.hy", line 3, in foo_m
    (defmacro foo-m [x](foo_s x))
  File "repro1.hy", line 1, in foo_s
    (eval-when-compile (defn foo-s [x] (+ x "str"))) ; << error here
TypeError: unsupported operand type(s) for +: 'Integer' and 'str'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/egx/miniforge3/bin/hy2py", line 12, in <module>
    sys.exit(hy2py_main())
...
```

I guess the change would to print the first traceback always, but the second when `HY_DEBUG=1` is set

**Reply by Kodiologist** [reply]
Cool, I could never figure out what `HY_DEBUG` was supposed to do.
